Add possibility to specify multiple disks in 'virt' module

dis-xcom opened the pull request:
Now disk profiles can contain multiple disks:

- backward compatible: if 'image' argument is set for virt.init()
  *and* 'image' attribute is missing for the first disk, then the
  first disk will use the 'image' from the method arguments

- multiple disks can be specified for a disk profile, see pillar
  example below

- if 'size' attribute set for qcow2-based image, then resulting
  image will be resized to the 'size' capacity (author: Ondrej Smola)

- empty disks allowed ('size' attribute is mandatory in this case)

Example disk profile 'three_disks':

    virt:
      disk:
        three_disks:
          - system:
              size: 4096
              image: http://path/to/ubuntu.qcow2
          - repository_snapshot:
              size: 8192
              image: /var/lib/libvirt/images/snapshot.qcow2
          - cinder-volume:
              size: 16384

### What does this PR do?

### What issues does this PR fix or reference?

### Previous Behavior
Remove this section if not relevant

### New Behavior
Remove this section if not relevant

### Tests written?

Yes/No

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/latest/topics/development/contributing.html) for best practices.

dis-xcom committed:
Add possibility to specify multiple disks in 'virt' module

Now disk profiles can contain multiple disks:

- backward compatible: if 'image' argument is set for virt.init()
  *and* 'image' attribute is missing for the first disk, then the
  first disk will use the 'image' from the method arguments

- multiple disks can be specified for a disk profile, see pillar
  example below

- if 'size' attribute set for qcow2-based image, then resulting
  image will be resized to the 'size' capacity (author: Ondrej Smola)

- empty disks allowed ('size' attribute is mandatory in this case)

Example disk profile 'three_disks':

virt:
  disk:
    three_disks:
      - system:
          size: 4096
          image: http://path/to/ubuntu.qcow2
      - repository_snapshot:
          size: 8192
          image: /var/lib/libvirt/images/snapshot.qcow2
      - cinder-volume:
          size: 16384

salt-jenkins commented:
@dis-xcom, thanks for your PR! By analyzing the history of the files in this pull request, we identified @techhat, @holmboe and @tgoodlet to be potential reviewers.

dis-xcom committed:
Fix debug message in virt.init

cachedout commented on salt/modules/virt.py:
```diff
@@ -673,6 +634,117 @@ def init(name,
+                try:
+                    os.makedirs(img_dir)
+                except OSError:
+                    pass
```
Are we sure we want to just ignore this?

dis-xcom replied:
This is the original code (just moved, see L620-623 in original file). I didn't analyze why this code was added early, so leave it as-is for now.

dis-xcom replied:
Here is that commit: https://github.com/saltstack/salt/commit/6413c11f2946c20676bae69f63d539e8ba44ebea#diff-b651f4b5ad5aad0bb9fd22bacd93caa1R593

tgoodlet commented on salt/modules/virt.py:
```diff
@@ -673,6 +634,117 @@ def init(name,
                     define_vol_xml_str(xml)
 
+            elif hypervisor in ['qemu', 'kvm']:
+
```
I'd say stick the ` for disk in diskp:` somewhere here; only loop when using the `libvirt` hypervisor.
Sticking lines 639-740 in a new function might make this a lot cleaner as well.

dis-xcom replied:
As I said - we should iterate 'diskp' for any type of hypervisors, the difference is: for qemu-kvm we create image, for other hypervisors we just describe the volume for libvirt (define_vol_xml_str() )

tgoodlet commented on salt/modules/virt.py:
```diff
@@ -574,96 +592,39 @@ def init(name,
-                    __salt__['cmd.run'](
-                        'qemu-img create -f qcow2 -o backing_file={} {}'
-                        .format(sfn, img_dest).split())
+        if not diskp[0][disk_name].get('image', None):
```
So if a user passes an `image` arg we only override it if the first entry in the profile **isn't defined**?
That seems like incorrect precedence; the arg should always take priority.

dis-xcom replied:
Default disk profile doesn't have 'image' attribute, so will always use 'image' from the module arguments.

But if a disk profile uses different images for different disks, there is no determined way to specify which images should be replaced with the image from argument:
disk_a:
    image: aaaa.qcow2
disk_b:
   image: bbbb.qcow2
disk_c:
   image: cccc.qcow2

If you think that the 'image' module argument should be for the first disk only, then it should be reflected at least in the docstring

tgoodlet commented on salt/modules/virt.py:
```diff
@@ -574,96 +592,39 @@ def init(name,
+            diskp[0][disk_name]['image'] = image
+
+    # Create multiple disks, empty or from specified images.
+    for disk in diskp:
```
Not following why this loop was moved here?
Do you need it if `hypervisor not in ['qemu', 'kvm']`?

I'm wondering if maybe it's worth moving all the 639-740 bit into a new function anyway; the indentation is getting a bit nested/convoluted.

dis-xcom replied:
Yes, we need it, see L665 in the original file (this code is reached if we have multiple-disks profile without specified images for non- qemu-kvm hypervisors).

There are two branches that can be nested in different orders:
- check for hypervisor and then iterate over disks for each hypervisor separatelly
- iterate over disks and check for hypervisor for creating the disks

The first option requires two different iteration loops in different places, the second option have a single iteration over disks in the disk profile. I think the second option will be more readable and easily to support.

As for moving the image creation code into a separated method - good idea, thanks!

goodboy reviewed (changes requested):
Some other notes:
- The legacy esxi/vmware stuff is probably never going to be added [(especially since `salt-cloud` already supports it)](https://docs.saltstack.com/en/latest/topics/cloud/vmware.html).
- the [libvirt backend for `salt-cloud`](https://github.com/saltstack/salt/pull/36709) might be a better place to stick this.
I know that I would have probably put my changes against that system if I had known about it at the time.

dis-xcom committed:
Move qemu-kvm image creation into a separate method

Also, changed priority for 'image' argument - now the module
argument will override the argument from disk profile.

dis-xcom committed:
Fix missing variable error

cachedout commented:
@tgoodlet How are you feeling about this one after the latest changes?

tgoodlet commented on salt/modules/virt.py:
```diff
@@ -580,90 +700,34 @@ def init(name,
+
+        for disk_name, args in six.iteritems(disk):
+
+            if hypervisor in ['esxi', 'vmware']:
```
@cachedout Can't we just drop this?

I mean is anyone actually going to add it given that `salt-cloud` is turning out to be the new system for all this?

dis-xcom replied:
I would prefer to make such changes in a different patch, because it could introduce regression that is not connected to the feature I'm trying to add :)

tgoodlet commented on salt/modules/virt.py:
```diff
@@ -353,6 +353,108 @@ def _qemu_image_info(path):
+            mask = os.umask(0)
+            os.umask(mask)
+
+            if disk_size and qcow2:
```
It seems odd to be the function needs to be called with something `False`y in order to avoid resizing? Maybe just set `disk_size=None`?

dis-xcom replied:
Just don't set 'size' attribute in the disk profile to avoid resizing, and it will be initialized with None in L744.

For standalone use of this method your comment makes sense, thanks!

tgoodlet commented on salt/modules/virt.py:
```diff
@@ -673,6 +737,49 @@ def init(name,
+                        .format(disk_file_name)
+                    )
+
+                img_dest = _qemu_image_create(
```
Yeah this is looking a lot easier to grok :)

dis-xcom committed:
Make _qemu_image_create() more independent
